src/Cardano/Wallet/Binary.hs

```diff
@@ -0,0 +1,374 @@
+decodeGenesisBlockHeader = do
+    _ <- CBOR.decodeListLenCanonicalOf 5
+    _ <- decodeProtocolMagic
+    previous <- decodePreviousBlockHeader
+    _ <- decodeGenesisProof
+    epochIndex <- decodeGenesisConsensusData
+    _ <- decodeGenesisExtraData
+    -- NOTE
+    -- Careful here, we do return a slot number of 0, which means that if we
+    -- naively parse all blocks from an epoch, two of them will have a slot
+    -- number of `0`. In practices, when parsing a full epoch, we can discard
+    -- the genesis block entirely and we won't bother about modelling this
+    -- extra complexity at the type-level. That's a bit dodgy though.
+    return $ BlockHeader epochIndex 0 previous
+
+decodeGenesisConsensusData :: CBOR.Decoder s Word64
+decodeGenesisConsensusData = do
+    _ <- CBOR.decodeListLenCanonicalOf 2
+    epochIndex <- CBOR.decodeWord64
+    _ <- decodeDifficulty
+    return epochIndex
+
+decodeGenesisExtraData :: CBOR.Decoder s ()
+decodeGenesisExtraData = do
+    _ <- CBOR.decodeListLenCanonicalOf 1
+    _ <- decodeAttributes
+    return ()
+
+decodeGenesisProof :: CBOR.Decoder s ()
+decodeGenesisProof = do
+    _ <- CBOR.decodeBytes -- Slot Leaders Hash
+    return ()
+
+decodeHeavyIndex :: CBOR.Decoder s ()
+decodeHeavyIndex = do
+    _ <- CBOR.decodeWord64 -- Epoch Index
+    return ()
+
+decodeLeaderKey :: CBOR.Decoder s ()
+decodeLeaderKey = do
+    _ <- CBOR.decodeBytes
+    return ()
+
+decodeLightIndex :: CBOR.Decoder s ()
+decodeLightIndex = do
+    _ <- CBOR.decodeListLenCanonicalOf 2
+    _ <- CBOR.decodeWord64 -- Epoch Index #1
+    _ <- CBOR.decodeWord64 -- Epoch Index #2
+    return ()
+
+decodeMainBlockBody :: CBOR.Decoder s (Set Tx)
+decodeMainBlockBody = do
+    _ <- CBOR.decodeListLenCanonicalOf 4
+    decodeTxPayload
+    -- NOTE:
+    -- Would remain after that:
+    --  - SscPayload
+    --  - DlsPayload
+    --  - UpdatePayload
+
+decodeMainBlockHeader :: CBOR.Decoder s BlockHeader
+decodeMainBlockHeader = do
+    _ <- CBOR.decodeListLenCanonicalOf 5
+    _ <- decodeProtocolMagic
+    previous <- decodePreviousBlockHeader
+    _ <- decodeMainProof
+    (epochIndex, slotNumber) <- decodeMainConsensusData
+    _ <- decodeMainExtraData
+    return $ BlockHeader epochIndex slotNumber previous
+
+decodeMainConsensusData :: CBOR.Decoder s (Word64, Word16)
+decodeMainConsensusData = do
+    _ <- CBOR.decodeListLenCanonicalOf 4
+    slot <- decodeSlotId
+    _ <- decodeLeaderKey
+    _ <- decodeDifficulty
+    _ <- decodeSignature
+    return slot
+
+decodeMainExtraData :: CBOR.Decoder s ()
+decodeMainExtraData = do
+    _ <- CBOR.decodeListLenCanonicalOf 4
+    _ <- decodeBlockVersion
+    _ <- decodeSoftwareVersion
+    _ <- decodeAttributes
+    _ <- decodeDataProof
+    return ()
+
+decodeMainProof :: CBOR.Decoder s ()
+decodeMainProof = do
+    CBOR.decodeListLenCanonicalOf 4
+    decodeTxProof
+    decodeMpcProof
+    decodeProxySKsProof
+    decodeUpdateProof
+
+decodeMpcProof :: CBOR.Decoder s ()
```

Comment:
Maybe here put a comment to say this stands for "multi-party-computation proof".

Reply:
Done.